### PR TITLE
Optimize documents fetching / filtering

### LINKED ISSUE #372: Optimize documents fetching / filtering
Our `fetch` methods read all documents before returning them to the user. That requires too much memory. Let's use channels.

Let's change our `fetch` methods to use this signature:

```go
// fetch fetches all documents from the given database and collection
// and sends them to the returned channel.
//
// The returned channel is always non-nil.
// The channel is closed when all documents are sent; the caller should always drain the channel.
// The returned error is the reason why the channel was closed:
// it is nil in case all documents were sent normally,
// error encountered during query initialization or iteration, or context's error.
//
// If the collection doesn't exist, fetch returns a closed channel and no error.
func (h *Handler) fetch(ctx context.Context, param sqlParam) (<-chan []*types.Document, error)
```

The implementation would use a goroutine that should handle context cancelation and always close the channel on exit. For now, let's use slices with lengths up to 2, and a channel with a capacity of 3. That ensures that all callers use `for` loops; we will change those numbers once we have cursor support and benchmarks.

Let's change that method for all handlers and all code that depends on it.

### PR DESCRIPTION
**Pull request by rumyantseva**
# Description

This PR closes #372.

## Readiness checklist

* [x] I set assignee, reviewers, labels, project and sprint.
* [x] I added tests for new functionality or bugfixes.
* [x] I ran `task all`, and it passed.
* [x] I added/updated documentation for exported and unexported functions, variables, types, etc.
* [x] I checked complex documentation rendering with `bin/godoc -index -play -timestamps -http=127.0.0.1:6060`.

### CODE REVIEW

File: internal/handlers/pg/pgdb/query_documents_test.go [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+			ctx, dbName, collectionName,
+			must.NotFail(types.NewDocument("id", fmt.Sprintf("%d", i))),
+		))
+	}
+
+	fetchedChan, err := pool.QueryDocuments(ctx, dbName, collectionName, "")
+	require.NoError(t, err)
+
+	<-fetchedChan
+	/*_, err = pool.Exec(ctx, "SELECT pg_terminate_backend(pg_backend_pid());")
```
Comment (rumyantseva) [inline internal/handlers/pg/pgdb/query_documents_test.go:159]: @AlekSi do you have any idea how to test this case? We wanted to emulate a problem with DB during iterations. Simple `pool.Close` doesn't work (it's hanging as the transaction is open), and also this `SELECT pg_terminate_backend(pg_backend_pid())` doesn't help either.

Comment (AlekSi) [reply internal/handlers/pg/pgdb/query_documents_test.go:159]: I don't know. Let's add TODO to think about it in the future

**Comment by rumyantseva** [comment]
For the `main` branch:
```
$ time task test-unit
task test-unit  15,10s user 3,98s system 202% cpu 9,419 total

$ time task test-integration-pg
task test-integration-pg  158,65s user 13,35s system 281% cpu 1:01,06 total

$ time task test-integration-mongodb
task test-integration-mongodb  4,98s user 1,96s system 98% cpu 7,068 total

$ time task test-integration-tigris
task test-integration-tigris  1,15s user 0,75s system 33% cpu 5,696 total
```

For the `rumyantseva:issue-372-optimize-fetching` branch:
```
$ time task test-unit
task test-unit  21,22s user 7,47s system 247% cpu 11,584 total

$ time task test-integration-pg

$ time task test-integration-mongodb
task test-integration-mongodb  4,77s user 1,95s system 106% cpu 6,307 total

$ time task test-integration-tigris
task test-integration-tigris  1,12s user 0,73s system 33% cpu 5,579 total

```

File: go.mod [hunk 2]
```diff
@@ -20,6 +20,8 @@ require (
 	google.golang.org/grpc v1.47.0
 )
 
+require golang.org/x/net v0.0.0-20220526153639-5463443f8c37
```
Comment (rumyantseva) [inline go.mod:23]: I don't understand why :/ (I don't think I import anything from `net`)

**Comment by seeforschauer** [comment]
tests are failing.

File: internal/handlers/pg/pgdb/query_documents_test.go [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+				assert.Equal(t, len(tc.docsPerIteration), iter)
+
+				return nil
+			})
+
+			require.NoError(t, err)
+		})
+	}
+
+	// Special case: cancel context before reading from channel.
```
Comment (seeforschauer) [inline internal/handlers/pg/pgdb/query_documents_test.go:122]: Let's remove unused code and extract it to the TODO.

File: internal/handlers/pg/pgdb/query_documents_test.go [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+	})*/
+
+	// Special case: querying a non-existing collection.
+	t.Run("non-existing_collection", func(t *testing.T) {
+		tx, err := pool.Begin(ctx)
+		require.NoError(t, err)
+
+		fetchedChan, err := pool.QueryDocuments(context.Background(), tx, dbName, collectionName+"_non-existing", "")
+		require.NoError(t, err)
+		doc, ok := <-fetchedChan
```
Comment (seeforschauer) [inline internal/handlers/pg/pgdb/query_documents_test.go:156]: Better name it not a doc but a `res` or smth similar, probably.

**Review by seeforschauer — COMMENTED** [review]
The review is not complete, just a few comments.

File: internal/handlers/pg/fetch.go [hunk 5]
```diff
@@ -31,27 +30,15 @@ type sqlParam struct {
...
-	return res, nil
+// If an error occurs before the fetching, the error is returned immediately.
+// The returned channel is always non-nil.
+// The channel is closed when all documents are sent; the caller should always drain the channel.
+// If an error occurs during fetching, the last message before closing the channel contains an error.
+// Context cancelation is not considered an error.
+//
+// If the collection doesn't exist, fetch returns a closed channel and no error.
+//
+func (h *Handler) fetch(ctx context.Context, querier pgxtype.Querier, param sqlParam) (<-chan pgdb.FetchedDocs, error) {
```
Comment (w84thesun) [inline internal/handlers/pg/fetch.go:42]: Do we need `fetch` anymore if all that it does now is call `QueryDocuments`?

File: internal/handlers/pg/msg_findandmodify.go [hunk 6]
```diff
@@ -56,28 +58,49 @@ func (h *Handler) MsgFindAndModify(ctx context.Context, msg *wire.OpMsg) (*wire.
 		return nil, err
 	}
 
-	fetchedDocs, err := h.fetch(ctx, params.sqlParam)
-	if err != nil {
-		return nil, err
-	}
+	// TODO: SortDocuments requires everything :(
```
Comment (w84thesun) [inline internal/handlers/pg/msg_findandmodify.go:61]: Can you please add more context here? Is it about the fact that we need to load everything to perform in-memory sort?

**Review by w84thesun — CHANGES_REQUESTED** [review]

File: internal/handlers/pg/fetch.go (hunk 5 above)
Comment (rumyantseva) [reply internal/handlers/pg/fetch.go]: Fixed

File: internal/handlers/pg/pgdb/query_documents_test.go (hunk 3 above)
Comment (rumyantseva) [reply internal/handlers/pg/pgdb/query_documents_test.go:121]: Yeah, that test actually works, I just forgot to uncomment it after some troubleshooting 🙈
Fixed!

File: internal/handlers/pg/pgdb/testutil_test.go [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+	"testing"
+
+	"github.com/stretchr/testify/require"
+	"go.uber.org/zap"
+
+	"github.com/FerretDB/FerretDB/internal/handlers/pg/pgdb"
+	"github.com/FerretDB/FerretDB/internal/util/testutil"
+)
+
+// Pool creates a new connection's connection getPool for testing.
```
Comment (seeforschauer) [inline internal/handlers/pg/pgdb/testutil_test.go:28]:
```suggestion
// getPool creates a new connection's connection getPool for testing.
```

File: internal/handlers/pg/pgdb/query_documents_test.go [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+
+	pool.DropDatabase(ctx, dbName)
+	require.NoError(t, pgdb.CreateDatabase(ctx, pool, dbName))
+
+	cases := []struct {
+		name             string
+		collection       string
+		documents        []*types.Document
+		docsPerIteration []int // how many documents should be fetched per each iteration
+		// len(docsPerIteration) is the amount of fetch iterations.
```
Comment (seeforschauer) [inline internal/handlers/pg/pgdb/query_documents_test.go:54]: let's remove it? Unused.

Comment (rumyantseva) [reply internal/handlers/pg/pgdb/query_documents_test.go]: @seeforschauer ok, I moved this part of documentation before the `docsPerIteration` definition. Hopefully, it makes more sense now.

**Review by seeforschauer — APPROVED** [review]

**Pull request merged**